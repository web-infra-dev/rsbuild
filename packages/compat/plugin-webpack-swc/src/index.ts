export { default as SwcLoader } from './loader';
export { pluginSwc } from './plugin';
export * from './binding';
export * from './types';
