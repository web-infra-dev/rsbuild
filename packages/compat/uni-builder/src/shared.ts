import { deepmerge } from '@rsbuild/shared/deepmerge';
import type {
  RsbuildPlugin,
  RsbuildConfig as RsbuildRspackConfig,
} from '@rsbuild/core';
import type { RsbuildConfig as RsbuildWebpackConfig } from '@rsbuild/webpack';
import type { UniBuilderRspackConfig, UniBuilderWebpackConfig } from './types';

export function parseCommonConfig<B = 'rspack' | 'webpack'>(
  uniBuilderConfig: B extends 'rspack'
    ? UniBuilderRspackConfig
    : UniBuilderWebpackConfig,
): {
  rsbuildConfig: B extends 'rspack'
    ? RsbuildRspackConfig
    : RsbuildWebpackConfig;
  rsbuildPlugins: RsbuildPlugin[];
} {
  const rsbuildConfig = deepmerge({}, uniBuilderConfig);
  const { output } = rsbuildConfig;

  if (output.cssModuleLocalIdentName) {
    output.cssModules ||= {};
    output.cssModules.localIdentName = output.cssModuleLocalIdentName;
    delete output.cssModuleLocalIdentName;
  }

  return {
    rsbuildConfig: rsbuildConfig,
    rsbuildPlugins: [],
  };
}
