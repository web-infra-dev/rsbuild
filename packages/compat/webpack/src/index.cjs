/**
 * CommonJS wrapper
 */
module.exports.webpackProvider = (...args) =>
  import('./index.js').then((i) => i.webpackProvider(...args));
