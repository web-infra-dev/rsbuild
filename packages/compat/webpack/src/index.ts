export { webpackProvider } from './provider';
export type { WebpackProvider } from './provider';
export { webpackBuild } from './core/build';
export { createDefaultConfig } from './config/defaults';
export type {
  LessLoaderOptions,
  SassLoaderOptions,
  PostCSSLoaderOptions,
  CSSLoaderOptions,
  StyleLoaderOptions,
} from '@rsbuild/shared';
export type {
  RsbuildPlugin,
  RsbuildPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  OutputConfig,
  SourceConfig,
  RsbuildConfig,
  SecurityConfig,
  NormalizedConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Hook Callback Types
  ModifyWebpackChainFn,
  ModifyWebpackConfigFn,
  ModifyWebpackChainUtils,
  ModifyWebpackConfigUtils,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSExtractOptions,
  HTMLPluginOptions,
} from './types';
