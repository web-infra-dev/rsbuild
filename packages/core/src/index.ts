/**
 * The methods and types exported from this file are considered as
 * the public API of @rsbuild/core.
 */
import { rspack } from '@rspack/core';
import type * as Rspack from '@rspack/core';
import * as __internalHelper from './internal';

// Core methods
export { loadEnv } from './loadEnv';
export { createRsbuild } from './createRsbuild';
export { loadConfig, defineConfig } from './config';

// Rsbuild version
export const version = RSBUILD_VERSION;

// Rspack instance
export { rspack };
export type { Rspack };

// Helpers
export { logger } from '@rsbuild/shared';
export { mergeRsbuildConfig } from './mergeConfig';

// Constants
export { PLUGIN_SWC_NAME, PLUGIN_CSS_NAME } from './constants';

// Types
export type {
  // Config Types
  RsbuildConfig,
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  SourceConfig,
  ServerConfig,
  OutputConfig,
  SecurityConfig,
  PerformanceConfig,
  ModuleFederationConfig,
  // Normalized Config Types
  NormalizedConfig,
  NormalizedDevConfig,
  NormalizedHtmlConfig,
  NormalizedToolsConfig,
  NormalizedSourceConfig,
  NormalizedServerConfig,
  NormalizedOutputConfig,
  NormalizedSecurityConfig,
  NormalizedPerformanceConfig,
  NormalizedModuleFederationConfig,
  // Plugin Types
  RsbuildPlugin,
  RsbuildPlugins,
  RsbuildPluginAPI,
} from './types';

export type {
  RsbuildMode,
  RsbuildEntry,
  RsbuildTarget,
  RsbuildContext,
  RsbuildInstance,
  CreateRsbuildOptions,
  InspectConfigOptions,
  // Subtypes of Config
  Minify,
  Polyfill,
  PrintUrls,
  PublicDir,
  Decorators,
  RspackRule,
  WatchFiles,
  CSSModules,
  CrossOrigin,
  ConsoleType,
  SplitChunks,
  BundlerChain,
  ClientConfig,
  ScriptInject,
  ChainedConfig,
  PostCSSPlugin,
  ScriptLoading,
  LegalComments,
  AliasStrategy,
  FilenameConfig,
  DistPathConfig,
  OutputStructure,
  ChainIdentifier,
  PreconnectOption,
  CSSLoaderOptions,
  ModifyChainUtils,
  StyleLoaderOptions,
  PostCSSLoaderOptions,
  ChainedConfigWithUtils,
  ModifyRspackConfigUtils,
  CSSModulesLocalsConvention,
  // Hook Callback Types
  OnExitFn,
  OnAfterBuildFn,
  OnAfterCreateCompilerFn,
  OnAfterStartDevServerFn,
  OnAfterStartProdServerFn,
  OnBeforeBuildFn,
  OnBeforeStartDevServerFn,
  OnBeforeStartProdServerFn,
  OnBeforeCreateCompilerFn,
  OnCloseDevServerFn,
  OnDevCompileDoneFn,
  ModifyBundlerChainFn,
  ModifyRspackConfigFn,
  ModifyRsbuildConfigFn,
  TransformFn,
  TransformHandler,
} from '@rsbuild/shared';

export {
  /**
   * @private
   */
  __internalHelper,
};
