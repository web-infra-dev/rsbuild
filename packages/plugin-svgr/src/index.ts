import path from 'node:path';
import {
  SVG_REGEX,
  deepmerge,
  getDistPath,
  getFilename,
  SCRIPT_REGEX,
} from '@rsbuild/shared';
import { PLUGIN_REACT_NAME } from '@rsbuild/plugin-react';
import type { RsbuildPlugin } from '@rsbuild/core';
import type { Config } from '@svgr/core';

export type SvgDefaultExport = 'component' | 'url';

export type PluginSvgrOptions = {
  /**
   * Configure SVGR options.
   * @see https://react-svgr.com/docs/options/
   */
  svgrOptions?: Config;

  /**
   * Whether to allow the use of default import and named import at the same time.
   * @default false
   */
  mixedImport?: boolean;
};

function getSvgoDefaultConfig() {
  return {
    plugins: [
      {
        name: 'preset-default',
        params: {
          overrides: {
            // viewBox is required to resize SVGs with CSS.
            // @see https://github.com/svg/svgo/issues/1128
            removeViewBox: false,
          },
        },
      },
      'prefixIds',
    ],
  };
}

export const pluginSvgr = (options: PluginSvgrOptions = {}): RsbuildPlugin => ({
  name: 'rsbuild:svgr',

  pre: [PLUGIN_REACT_NAME],

  setup(api) {
    api.modifyBundlerChain(async (chain, { isProd, CHAIN_ID }) => {
      const config = api.getNormalizedConfig();
      const distDir = getDistPath(config, 'svg');
      const filename = getFilename(config, 'svg', isProd);
      const outputName = path.posix.join(distDir, filename);
      const { dataUriLimit } = config.output;
      const maxSize =
        typeof dataUriLimit === 'number' ? dataUriLimit : dataUriLimit.svg;

      // delete Rsbuild builtin SVG rules
      chain.module.rules.delete(CHAIN_ID.RULE.SVG);

      const rule = chain.module.rule(CHAIN_ID.RULE.SVG).test(SVG_REGEX);

      const svgrOptions = deepmerge(
        {
          svgo: true,
          svgoConfig: getSvgoDefaultConfig(),
        },
        options.svgrOptions || {},
      );

      // force to url: "foo.svg?url",
      rule
        .oneOf(CHAIN_ID.ONE_OF.SVG_URL)
        .type('asset/resource')
        .resourceQuery(/(__inline=false|url)/)
        .set('generator', {
          filename: outputName,
        });

      // force to inline: "foo.svg?inline"
      rule
        .oneOf(CHAIN_ID.ONE_OF.SVG_INLINE)
        .type('asset/inline')
        .resourceQuery(/inline/);

      // force to react component: "foo.svg?react"
      rule
        .oneOf(CHAIN_ID.ONE_OF.SVG_REACT)
        .type('javascript/auto')
        .resourceQuery(/react/)
        .use(CHAIN_ID.USE.SVGR)
        .loader(path.resolve(__dirname, './loader'))
        .options({
          ...svgrOptions,
          exportType: 'default',
        } satisfies Config)
        .end();

      // SVG in JS files
      const { mixedImport = false } = options;
      if (mixedImport || svgrOptions.exportType) {
        const { exportType = mixedImport ? 'named' : undefined } = svgrOptions;

        const svgRule = rule
          .oneOf(CHAIN_ID.ONE_OF.SVG)
          .type('javascript/auto')
          // The issuer option ensures that SVGR will only apply if the SVG is imported from a JS file.
          .set('issuer', [SCRIPT_REGEX])
          .use(CHAIN_ID.USE.SVGR)
          .loader(path.resolve(__dirname, './loader'))
          .options({
            ...svgrOptions,
            exportType,
          })
          .end();

        /**
         * For mixed import.
         * @example import logoUrl, { ReactComponent } from './logo.svg';`
         */
        if (mixedImport && exportType === 'named') {
          svgRule
            .use(CHAIN_ID.USE.URL)
            .loader(path.join(__dirname, '../compiled', 'url-loader'))
            .options({
              limit: maxSize,
              name: outputName,
            });
        }
      }

      // SVG as assets
      rule
        .oneOf(CHAIN_ID.ONE_OF.SVG_ASSET)
        .type('asset')
        .parser({
          // Inline SVG if size < maxSize
          dataUrlCondition: {
            maxSize,
          },
        })
        .set('generator', {
          filename: outputName,
        });

      // apply current JS transform rule to SVGR rules
      const jsRule = chain.module.rules.get(CHAIN_ID.RULE.JS);

      [CHAIN_ID.USE.SWC, CHAIN_ID.USE.BABEL].some((jsUseId) => {
        const use = jsRule.uses.get(jsUseId);

        if (!use) {
          return false;
        }

        [CHAIN_ID.ONE_OF.SVG, CHAIN_ID.ONE_OF.SVG_REACT].forEach((oneOfId) => {
          if (!rule.oneOfs.has(oneOfId)) {
            return;
          }

          rule
            .oneOf(oneOfId)
            .use(jsUseId)
            .before(CHAIN_ID.USE.SVGR)
            .loader(use.get('loader'))
            .options(use.get('options'));
        });

        return true;
      });
    });
  },
});
