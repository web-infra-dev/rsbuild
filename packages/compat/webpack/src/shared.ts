import { __internalHelper } from '@rsbuild/core';

const {
  getChainUtils,
  initRsbuildConfig,
  createDevServer,
  formatStats,
  getDevMiddleware,
  getStatsOptions,
  stringifyConfig,
  outputInspectConfigFiles,
} = __internalHelper;

export {
  getChainUtils,
  initRsbuildConfig,
  createDevServer,
  formatStats,
  getDevMiddleware,
  getStatsOptions,
  stringifyConfig,
  outputInspectConfigFiles,
};

export type InternalContext = __internalHelper.InternalContext;
