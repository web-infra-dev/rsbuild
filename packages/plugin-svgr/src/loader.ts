/**
 * Copyright (c) 2017, Smooth Code
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * modified from https://github.com/gregberge/svgr/blob/7595d378b73d4826a4cead165b3f32386b07315b/packages/webpack/src/index.ts
 */

import { normalize } from 'node:path';
import { callbackify } from 'node:util';
import type { Rspack } from '@rsbuild/core';
import { type Config, type State, transform } from '@svgr/core';
import jsx from '@svgr/plugin-jsx';
import svgo from '@svgr/plugin-svgo';

const transformSvg = callbackify(
  async (contents: string, config: Config, state: Partial<State>) =>
    transform(contents, config, state),
);

function svgrLoader(
  this: Rspack.LoaderContext<Config>,
  contents: string,
): void {
  this?.cacheable();

  const callback = this.async();

  const options = this.getOptions();

  const previousExport = (() => {
    if (contents.startsWith('export ')) return contents;
    const exportMatches = contents.match(/^module.exports\s*=\s*(.*)/);
    return exportMatches ? `export default ${exportMatches[1]}` : null;
  })();

  const state = {
    caller: {
      name: '@rsbuild/plugin-svgr',
      previousExport,
      defaultPlugins: [svgo, jsx],
    },
    filePath: normalize(this.resourcePath),
  };

  if (!previousExport) {
    transformSvg(contents, options, state, callback);
  } else {
    this.fs.readFile(this.resourcePath, (err: Error, result: unknown) => {
      if (err) {
        callback(err);
        return;
      }
      transformSvg(String(result), options, state, (err, content) => {
        if (err) {
          callback(err);
          return;
        }
        callback(null, content);
      });
    });
  }
}

export default svgrLoader;
