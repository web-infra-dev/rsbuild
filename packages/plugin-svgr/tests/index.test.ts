import { expect, describe, it } from 'vitest';
import { type PluginSvgrOptions, pluginSvgr } from '../src';
import { createStubRsbuild } from '@scripts/test-helper';
import { pluginReact } from '@rsbuild/plugin-react';
import { SVG_REGEX } from '@rsbuild/shared';

describe('svgr', () => {
  const cases: Array<{ name: string; pluginConfig: PluginSvgrOptions }> = [
    {
      name: 'exportType named / mixedImport true',
      pluginConfig: {
        svgrOptions: {
          exportType: 'named',
        },
        mixedImport: true,
      },
    },
    {
      name: 'exportType named / mixedImport false',
      pluginConfig: {
        svgrOptions: {
          exportType: 'named',
        },
        mixedImport: false,
      },
    },
    {
      name: 'exportType default / mixedImport false',
      pluginConfig: {
        svgrOptions: {
          exportType: 'default',
        },
        mixedImport: false,
      },
    },
    {
      name: 'exportType default / mixedImport true',
      pluginConfig: {
        svgrOptions: {
          exportType: 'default',
        },
        mixedImport: true,
      },
    },
    {
      name: 'configure SVGR options',
      pluginConfig: {
        svgrOptions: {
          svgoConfig: {
            datauri: 'base64',
          },
        },
      },
    },
  ];

  it.each(cases)('$name', async (item) => {
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {},
    });

    rsbuild.addPlugins([pluginSvgr(item.pluginConfig), pluginReact()]);

    const config = await rsbuild.unwrapConfig();

    expect(
      config.module.rules.find((r) => r.test === SVG_REGEX),
    ).toMatchSnapshot();
  });
});
