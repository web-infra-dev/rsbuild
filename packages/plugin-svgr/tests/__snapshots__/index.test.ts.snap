// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`svgr > 'configure SVGR options' 1`] = `
{
  "oneOf": [
    {
      "generator": {
        "filename": "static/svg/[name].[contenthash:8].svg",
      },
      "resourceQuery": /\\(__inline=false\\|url\\)/,
      "type": "asset/resource",
    },
    {
      "resourceQuery": /inline/,
      "type": "asset/inline",
    },
    {
      "resourceQuery": /react/,
      "type": "javascript/auto",
      "use": [
        {
          "loader": "builtin:swc-loader",
          "options": {
            "env": {
              "coreJs": "3.36",
              "mode": "usage",
              "shippedProposals": true,
              "targets": [
                "chrome >= 87",
                "edge >= 88",
                "firefox >= 78",
                "safari >= 14",
              ],
            },
            "isModule": "unknown",
            "jsc": {
              "externalHelpers": true,
              "parser": {
                "decorators": true,
                "syntax": "typescript",
                "tsx": true,
              },
              "preserveAllComments": true,
              "transform": {
                "decoratorMetadata": true,
                "legacyDecorator": true,
                "react": {
                  "development": false,
                  "refresh": true,
                  "runtime": "automatic",
                },
                "useDefineForClassFields": false,
              },
            },
            "sourceMaps": true,
          },
        },
        {
          "loader": "<ROOT>/packages/plugin-svgr/src/loader",
          "options": {
            "exportType": "default",
            "svgo": true,
            "svgoConfig": {
              "datauri": "base64",
              "plugins": [
                {
                  "name": "preset-default",
                  "params": {
                    "overrides": {
                      "removeViewBox": false,
                    },
                  },
                },
                "prefixIds",
              ],
            },
          },
        },
      ],
    },
    {
      "generator": {
        "filename": "static/svg/[name].[contenthash:8].svg",
      },
      "parser": {
        "dataUrlCondition": {
          "maxSize": 10000,
        },
      },
      "type": "asset",
    },
  ],
  "test": /\\\\\\.svg\\$/,
}
`;

exports[`svgr > 'exportType default / mixedImport false' 1`] = `
{
  "oneOf": [
    {
      "generator": {
        "filename": "static/svg/[name].[contenthash:8].svg",
      },
      "resourceQuery": /\\(__inline=false\\|url\\)/,
      "type": "asset/resource",
    },
    {
      "resourceQuery": /inline/,
      "type": "asset/inline",
    },
    {
      "resourceQuery": /react/,
      "type": "javascript/auto",
      "use": [
        {
          "loader": "builtin:swc-loader",
          "options": {
            "env": {
              "coreJs": "3.36",
              "mode": "usage",
              "shippedProposals": true,
              "targets": [
                "chrome >= 87",
                "edge >= 88",
                "firefox >= 78",
                "safari >= 14",
              ],
            },
            "isModule": "unknown",
            "jsc": {
              "externalHelpers": true,
              "parser": {
                "decorators": true,
                "syntax": "typescript",
                "tsx": true,
              },
              "preserveAllComments": true,
              "transform": {
                "decoratorMetadata": true,
                "legacyDecorator": true,
                "react": {
                  "development": false,
                  "refresh": true,
                  "runtime": "automatic",
                },
                "useDefineForClassFields": false,
              },
            },
            "sourceMaps": true,
          },
        },
        {
          "loader": "<ROOT>/packages/plugin-svgr/src/loader",
          "options": {
            "exportType": "default",
            "svgo": true,
            "svgoConfig": {
              "plugins": [
                {
                  "name": "preset-default",
                  "params": {
                    "overrides": {
                      "removeViewBox": false,
                    },
                  },
                },
                "prefixIds",
              ],
            },
          },
        },
      ],
    },
    {
      "issuer": [
        /\\\\\\.\\(\\?:js\\|jsx\\|mjs\\|cjs\\|ts\\|tsx\\|mts\\|cts\\)\\$/,
      ],
      "type": "javascript/auto",
      "use": [
        {
          "loader": "builtin:swc-loader",
          "options": {
            "env": {
              "coreJs": "3.36",
              "mode": "usage",
              "shippedProposals": true,
              "targets": [
                "chrome >= 87",
                "edge >= 88",
                "firefox >= 78",
                "safari >= 14",
              ],
            },
            "isModule": "unknown",
            "jsc": {
              "externalHelpers": true,
              "parser": {
                "decorators": true,
                "syntax": "typescript",
                "tsx": true,
              },
              "preserveAllComments": true,
              "transform": {
                "decoratorMetadata": true,
                "legacyDecorator": true,
                "react": {
                  "development": false,
                  "refresh": true,
                  "runtime": "automatic",
                },
                "useDefineForClassFields": false,
              },
            },
            "sourceMaps": true,
          },
        },
        {
          "loader": "<ROOT>/packages/plugin-svgr/src/loader",
          "options": {
            "exportType": "default",
            "svgo": true,
            "svgoConfig": {
              "plugins": [
                {
                  "name": "preset-default",
                  "params": {
                    "overrides": {
                      "removeViewBox": false,
                    },
                  },
                },
                "prefixIds",
              ],
            },
          },
        },
      ],
    },
    {
      "generator": {
        "filename": "static/svg/[name].[contenthash:8].svg",
      },
      "parser": {
        "dataUrlCondition": {
          "maxSize": 10000,
        },
      },
      "type": "asset",
    },
  ],
  "test": /\\\\\\.svg\\$/,
}
`;

exports[`svgr > 'exportType default / mixedImport true' 1`] = `
{
  "oneOf": [
    {
      "generator": {
        "filename": "static/svg/[name].[contenthash:8].svg",
      },
      "resourceQuery": /\\(__inline=false\\|url\\)/,
      "type": "asset/resource",
    },
    {
      "resourceQuery": /inline/,
      "type": "asset/inline",
    },
    {
      "resourceQuery": /react/,
      "type": "javascript/auto",
      "use": [
        {
          "loader": "builtin:swc-loader",
          "options": {
            "env": {
              "coreJs": "3.36",
              "mode": "usage",
              "shippedProposals": true,
              "targets": [
                "chrome >= 87",
                "edge >= 88",
                "firefox >= 78",
                "safari >= 14",
              ],
            },
            "isModule": "unknown",
            "jsc": {
              "externalHelpers": true,
              "parser": {
                "decorators": true,
                "syntax": "typescript",
                "tsx": true,
              },
              "preserveAllComments": true,
              "transform": {
                "decoratorMetadata": true,
                "legacyDecorator": true,
                "react": {
                  "development": false,
                  "refresh": true,
                  "runtime": "automatic",
                },
                "useDefineForClassFields": false,
              },
            },
            "sourceMaps": true,
          },
        },
        {
          "loader": "<ROOT>/packages/plugin-svgr/src/loader",
          "options": {
            "exportType": "default",
            "svgo": true,
            "svgoConfig": {
              "plugins": [
                {
                  "name": "preset-default",
                  "params": {
                    "overrides": {
                      "removeViewBox": false,
                    },
                  },
                },
                "prefixIds",
              ],
            },
          },
        },
      ],
    },
    {
      "issuer": [
        /\\\\\\.\\(\\?:js\\|jsx\\|mjs\\|cjs\\|ts\\|tsx\\|mts\\|cts\\)\\$/,
      ],
      "type": "javascript/auto",
      "use": [
        {
          "loader": "builtin:swc-loader",
          "options": {
            "env": {
              "coreJs": "3.36",
              "mode": "usage",
              "shippedProposals": true,
              "targets": [
                "chrome >= 87",
                "edge >= 88",
                "firefox >= 78",
                "safari >= 14",
              ],
            },
            "isModule": "unknown",
            "jsc": {
              "externalHelpers": true,
              "parser": {
                "decorators": true,
                "syntax": "typescript",
                "tsx": true,
              },
              "preserveAllComments": true,
              "transform": {
                "decoratorMetadata": true,
                "legacyDecorator": true,
                "react": {
                  "development": false,
                  "refresh": true,
                  "runtime": "automatic",
                },
                "useDefineForClassFields": false,
              },
            },
            "sourceMaps": true,
          },
        },
        {
          "loader": "<ROOT>/packages/plugin-svgr/src/loader",
          "options": {
            "exportType": "default",
            "svgo": true,
            "svgoConfig": {
              "plugins": [
                {
                  "name": "preset-default",
                  "params": {
                    "overrides": {
                      "removeViewBox": false,
                    },
                  },
                },
                "prefixIds",
              ],
            },
          },
        },
      ],
    },
    {
      "generator": {
        "filename": "static/svg/[name].[contenthash:8].svg",
      },
      "parser": {
        "dataUrlCondition": {
          "maxSize": 10000,
        },
      },
      "type": "asset",
    },
  ],
  "test": /\\\\\\.svg\\$/,
}
`;

exports[`svgr > 'exportType named / mixedImport false' 1`] = `
{
  "oneOf": [
    {
      "generator": {
        "filename": "static/svg/[name].[contenthash:8].svg",
      },
      "resourceQuery": /\\(__inline=false\\|url\\)/,
      "type": "asset/resource",
    },
    {
      "resourceQuery": /inline/,
      "type": "asset/inline",
    },
    {
      "resourceQuery": /react/,
      "type": "javascript/auto",
      "use": [
        {
          "loader": "builtin:swc-loader",
          "options": {
            "env": {
              "coreJs": "3.36",
              "mode": "usage",
              "shippedProposals": true,
              "targets": [
                "chrome >= 87",
                "edge >= 88",
                "firefox >= 78",
                "safari >= 14",
              ],
            },
            "isModule": "unknown",
            "jsc": {
              "externalHelpers": true,
              "parser": {
                "decorators": true,
                "syntax": "typescript",
                "tsx": true,
              },
              "preserveAllComments": true,
              "transform": {
                "decoratorMetadata": true,
                "legacyDecorator": true,
                "react": {
                  "development": false,
                  "refresh": true,
                  "runtime": "automatic",
                },
                "useDefineForClassFields": false,
              },
            },
            "sourceMaps": true,
          },
        },
        {
          "loader": "<ROOT>/packages/plugin-svgr/src/loader",
          "options": {
            "exportType": "default",
            "svgo": true,
            "svgoConfig": {
              "plugins": [
                {
                  "name": "preset-default",
                  "params": {
                    "overrides": {
                      "removeViewBox": false,
                    },
                  },
                },
                "prefixIds",
              ],
            },
          },
        },
      ],
    },
    {
      "issuer": [
        /\\\\\\.\\(\\?:js\\|jsx\\|mjs\\|cjs\\|ts\\|tsx\\|mts\\|cts\\)\\$/,
      ],
      "type": "javascript/auto",
      "use": [
        {
          "loader": "builtin:swc-loader",
          "options": {
            "env": {
              "coreJs": "3.36",
              "mode": "usage",
              "shippedProposals": true,
              "targets": [
                "chrome >= 87",
                "edge >= 88",
                "firefox >= 78",
                "safari >= 14",
              ],
            },
            "isModule": "unknown",
            "jsc": {
              "externalHelpers": true,
              "parser": {
                "decorators": true,
                "syntax": "typescript",
                "tsx": true,
              },
              "preserveAllComments": true,
              "transform": {
                "decoratorMetadata": true,
                "legacyDecorator": true,
                "react": {
                  "development": false,
                  "refresh": true,
                  "runtime": "automatic",
                },
                "useDefineForClassFields": false,
              },
            },
            "sourceMaps": true,
          },
        },
        {
          "loader": "<ROOT>/packages/plugin-svgr/src/loader",
          "options": {
            "exportType": "named",
            "svgo": true,
            "svgoConfig": {
              "plugins": [
                {
                  "name": "preset-default",
                  "params": {
                    "overrides": {
                      "removeViewBox": false,
                    },
                  },
                },
                "prefixIds",
              ],
            },
          },
        },
      ],
    },
    {
      "generator": {
        "filename": "static/svg/[name].[contenthash:8].svg",
      },
      "parser": {
        "dataUrlCondition": {
          "maxSize": 10000,
        },
      },
      "type": "asset",
    },
  ],
  "test": /\\\\\\.svg\\$/,
}
`;

exports[`svgr > 'exportType named / mixedImport true' 1`] = `
{
  "oneOf": [
    {
      "generator": {
        "filename": "static/svg/[name].[contenthash:8].svg",
      },
      "resourceQuery": /\\(__inline=false\\|url\\)/,
      "type": "asset/resource",
    },
    {
      "resourceQuery": /inline/,
      "type": "asset/inline",
    },
    {
      "resourceQuery": /react/,
      "type": "javascript/auto",
      "use": [
        {
          "loader": "builtin:swc-loader",
          "options": {
            "env": {
              "coreJs": "3.36",
              "mode": "usage",
              "shippedProposals": true,
              "targets": [
                "chrome >= 87",
                "edge >= 88",
                "firefox >= 78",
                "safari >= 14",
              ],
            },
            "isModule": "unknown",
            "jsc": {
              "externalHelpers": true,
              "parser": {
                "decorators": true,
                "syntax": "typescript",
                "tsx": true,
              },
              "preserveAllComments": true,
              "transform": {
                "decoratorMetadata": true,
                "legacyDecorator": true,
                "react": {
                  "development": false,
                  "refresh": true,
                  "runtime": "automatic",
                },
                "useDefineForClassFields": false,
              },
            },
            "sourceMaps": true,
          },
        },
        {
          "loader": "<ROOT>/packages/plugin-svgr/src/loader",
          "options": {
            "exportType": "default",
            "svgo": true,
            "svgoConfig": {
              "plugins": [
                {
                  "name": "preset-default",
                  "params": {
                    "overrides": {
                      "removeViewBox": false,
                    },
                  },
                },
                "prefixIds",
              ],
            },
          },
        },
      ],
    },
    {
      "issuer": [
        /\\\\\\.\\(\\?:js\\|jsx\\|mjs\\|cjs\\|ts\\|tsx\\|mts\\|cts\\)\\$/,
      ],
      "type": "javascript/auto",
      "use": [
        {
          "loader": "builtin:swc-loader",
          "options": {
            "env": {
              "coreJs": "3.36",
              "mode": "usage",
              "shippedProposals": true,
              "targets": [
                "chrome >= 87",
                "edge >= 88",
                "firefox >= 78",
                "safari >= 14",
              ],
            },
            "isModule": "unknown",
            "jsc": {
              "externalHelpers": true,
              "parser": {
                "decorators": true,
                "syntax": "typescript",
                "tsx": true,
              },
              "preserveAllComments": true,
              "transform": {
                "decoratorMetadata": true,
                "legacyDecorator": true,
                "react": {
                  "development": false,
                  "refresh": true,
                  "runtime": "automatic",
                },
                "useDefineForClassFields": false,
              },
            },
            "sourceMaps": true,
          },
        },
        {
          "loader": "<ROOT>/packages/plugin-svgr/src/loader",
          "options": {
            "exportType": "named",
            "svgo": true,
            "svgoConfig": {
              "plugins": [
                {
                  "name": "preset-default",
                  "params": {
                    "overrides": {
                      "removeViewBox": false,
                    },
                  },
                },
                "prefixIds",
              ],
            },
          },
        },
        {
          "loader": "<ROOT>/packages/plugin-svgr/compiled/url-loader",
          "options": {
            "limit": 10000,
            "name": "static/svg/[name].[contenthash:8].svg",
          },
        },
      ],
    },
    {
      "generator": {
        "filename": "static/svg/[name].[contenthash:8].svg",
      },
      "parser": {
        "dataUrlCondition": {
          "maxSize": 10000,
        },
      },
      "type": "asset",
    },
  ],
  "test": /\\\\\\.svg\\$/,
}
`;
